### PR TITLE
Improvements to M3 demo app

### PR DESCRIPTION
**Pull request by guidezpl**
## Description

This PR includes these changes:
* Added an integration test to check that the app can run without runtime errors (prevent https://github.com/flutter/samples/issues/1642 from happening in the future)
* Remove a duplicate `MaterialApp`. In order to do this, refactor the large `Material3App` widget into `App` and `Home`. Shared constants were moved to `constants.dart`.
* Fix `Focus` handling for the app, in particular, to make it easier to focus on elements (b/267773244). Focus now starts logically with navigation and options, then onto components, grouped by list, then by component group. A user can tap in a component card to move the focus there rather than tab all the way through the app. This change was tested manually.

## Pre-launch Checklist

- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I read the [Contributors Guide].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-devrel channel on [Discord].

[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Contributors Guide]: https://github.com/flutter/samples/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: experimental/material_3_demo/lib/constants.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2021 The Flutter team. All rights reserved.
```
Comment (guidezpl) [inline experimental/material_3_demo/lib/constants.dart:1]: No changes, simply moved from `main.dart`

File: experimental/material_3_demo/lib/home.dart [hunk 2]
```diff
@@ -0,0 +1,665 @@
+// Copyright 2021 The Flutter team. All rights reserved.
```
Comment (guidezpl) [inline experimental/material_3_demo/lib/home.dart:1]: Factored out of `main.dart`

File: material_3_demo/integration_test/integration_test.dart [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2021 The Flutter team. All rights reserved.
```
Comment (guidezpl) [inline material_3_demo/integration_test/integration_test.dart:1]: All changes below are the same as the changes to `experimental/material_3_demo`

**Review by domesticmouse — APPROVED** [review]
LGTM once @QuncCccccc signs off

**Review by QuncCccccc — COMMENTED** [review]
Looks good! Just left one question. Thanks!

File: experimental/material_3_demo/lib/component_screen.dart [hunk 4]
```diff
@@ -1042,23 +1048,26 @@ class _NavigationBarsState extends State<NavigationBars> {
...
-        if (!widget.isExampleBar) widget.onSelectItem!(index);
-      },
-      destinations: widget.isExampleBar && isBadgeExample
-          ? barWithBadgeDestinations
-          : widget.isExampleBar
-              ? exampleBarDestinations
-              : appBarDestinations,
+    // App NavigationBar should get first focus.
+    Widget navigationBar = Focus(
+      autofocus: !widget.isBadgeExample,
```
[the remark this answers is not in the document]
Comment (guidezpl) [reply experimental/material_3_demo/lib/component_screen.dart:1053]: Indeed, fixed!

**Review by QuncCccccc — APPROVED** [review]
LGTM. Will deploy this on web later. Thanks!

**Pull request merged**